lib/LowerAddrSpaceCastPass.cpp

```diff
@@ -350,20 +350,92 @@ Value *clspv::LowerAddrSpaceCastPass::visitPtrToIntInst(PtrToIntInst &I) {
   return ptrToInt;
 }
 
+static bool DependsOnPhiNode(const Value *Value, const PHINode *PhiNode,
+                             DenseSet<const llvm::Value *> &Visited) {
+  if (!isa<Instruction>(Value)) {
+    return false;
+  }
+  if (Visited.contains(Value)) {
+    return false;
+  }
+
+  Visited.insert(Value);
+
+  // If the value is the PHI node itself, return true
+  if (Value == PhiNode) {
+    return true;
+  }
+
+  // Recursively check if any of the operands depend on the PHI node
+  for (const auto *Op : cast<Instruction>(Value)->operand_values()) {
+    if (DependsOnPhiNode(Op, PhiNode, Visited)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
 Value *clspv::LowerAddrSpaceCastPass::visitPHINode(llvm::PHINode &I) {
+  if (!isGenericPTy(I.getType())) {
+    // Nothing to do.
+    return &I;
+  }
+
   IRBuilder<> B(&I);
-  // NOTE: We assume that the first incoming value does not depend on I.
-  auto *V1 = visit(I.getIncomingValue(0));
-  auto *B1 = I.getIncomingBlock(0);
 
-  // Register the replacement early in order to avoid recursive calls
-  // when an incoming value depends on this PHI node.
-  auto Phi = B.CreatePHI(V1->getType(), I.getNumIncomingValues());
+  unsigned N = I.getNumIncomingValues();
+
+  // Analyse the incoming values anche check whether the types agree.
+  // Ignore any incoming value that depends on the node itself, as that would
+  // lead to infinite recursion. Instead, we delay processing them until
+  // after we have registered the replacement.
+  SmallVector<Value *, 2> Replacements(N);
+
+  DenseSet<const llvm::Value *> VisitedForSelfDep;
+  SmallVector<unsigned, 2> DependentOperands;
+  Type *CommonTy = nullptr;
+  bool HasCommonTy = true;
+
+  for (unsigned j = 0; j < N; ++j) {
+    VisitedForSelfDep.clear();
+    auto *V = I.getIncomingValue(j);
+
+    if (DependsOnPhiNode(V, &I, VisitedForSelfDep)) {
+      DependentOperands.push_back(j);
+    } else {
+      V = visit(V);
+      if (HasCommonTy && !CommonTy) {
+        CommonTy = V->getType();
+      }
+      HasCommonTy &= CommonTy == V->getType();
+    }
+    Replacements[j] = V;
+  }
+
+  if (!HasCommonTy) {
+    // We don't have a common address space. To keep the node legal, we have to
+    // add addrspacecasts again.
+    CommonTy = B.getPtrTy(clspv::AddressSpace::Generic);
+    for (unsigned j = 0; j < N; ++j) {
+      if (Replacements[j]->getType() != CommonTy) {
+        Replacements[j] = B.CreateAddrSpaceCast(Replacements[j], CommonTy);
+      }
+    }
+  }
+
+  auto *Phi = B.CreatePHI(CommonTy, N);
+
```

Comment:
The goal is to create a phi that is not using generic addrspace. I don't understand what's happening here then.

Reply:
So, if the phi types have been resolved to different address spaces (the code has a conditional assignment from different address spaces), we can't fix this phi node at the moment. So they are added back to keep the node valid, but we know that generic phi nodes after this pass are there because of these conditional assignments.

The reason I did this is because I'm working on a pass that will try to fix them later on (I'll post a draft PR soon :) )

Reply:
The goal of this pass is to remove generic addrspace. Could this be implemented in that pass then?

Reply:
It can't because this pass would crash, but I've removed this for now. I'll post a new implementation in a separate PR